### PR TITLE
bugfix: mount etcd key, cert and CA separately

### LINKED ISSUE #1433: Can not use same file for `etcdKeyFile` & `etcdCertFile`
## Version

`1.0.0`

## Example `/etc/calico/calicoctl.yml` file

```
apiVersion: v1
kind: calicoApiConfig
metadata:
spec:
  datastoreType: etcdv2
  etcdEndpoints: https://etcd1:2379,https://etcd2:2379,https://etcd3:2379
  etcdKeyFile: /etc/tls/etcdclientpem
  etcdCertFile: /etc/tls/etcdclientpem
  etcdCACertFile: /etc/tls/etcdca
```
Where `/etc/tls/etcdclientpem` contains both tls `key` and `cert`.

## Command to reproduce
`sudo calicoctl node run --config /etc/calico/calicoctl.yml`

## Actual docker command triggered

`docker run --net=host --privileged --name=calico-node -d --restart=always -e ETCD_AUTHORITY= -e ETCD_CA_CERT_FILE=/etc/calico/certs/ca_cert.crt -e CALICO_NETWORKING_BACKEND=bird -e CALICO_LIBNETWORK_ENABLED=true -e CALICO_LIBNETWORK_IFPREFIX=cali -e ETCD_SCHEME= -e ETCD_KEY_FILE=/etc/calico/certs/key.pem -e ETCD_CERT_FILE=/etc/calico/certs/cert.crt -e NODENAME=swarm-node-0 -e NO_DEFAULT_POOLS= -e ETCD_ENDPOINTS=https://etcd1:2379,https://etcd2:2379,https://etcd3:2379 -v /etc/tls/etcdclientpem:/etc/calico/certs/cert.crt -v /var/run/calico:/var/run/calico -v /lib/modules:/lib/modules -v /var/log/calico:/var/log/calico -v /run/docker/plugins:/run/docker/plugins -v /var/run/docker.sock:/var/run/docker.sock -v /etc/tls/etcdca:/etc/calico/certs/ca_cert.crt calico/node:v1.0.0`

## Expected docker command

`docker run --net=host --privileged --name=calico-node -d --restart=always -e NO_DEFAULT_POOLS= -e CALICO_LIBNETWORK_ENABLED=true -e CALICO_LIBNETWORK_IFPREFIX=cali -e ETCD_ENDPOINTS=https://etcd1:2379,https://etcd2:2379,https://etcd3:2379 -e ETCD_AUTHORITY= -e ETCD_CA_CERT_FILE=/etc/calico/certs/ca_cert.crt -e NODENAME=swarm-node-0 -e CALICO_NETWORKING_BACKEND=bird -e ETCD_CERT_FILE=/etc/calico/certs/cert.crt -e ETCD_SCHEME= -e ETCD_KEY_FILE=/etc/calico/certs/key.pem -v /etc/tls/etcdca:/etc/calico/certs/ca_cert.crt -v /etc/tls/etcdclientpem:/etc/calico/certs/key.pem -v /etc/tls/etcdclientpem:/etc/calico/certs/cert.crt -v /var/run/calico:/var/run/calico -v /lib/modules:/lib/modules -v /var/log/calico:/var/log/calico -v /run/docker/plugins:/run/docker/plugins -v /var/run/docker.sock:/var/run/docker.sock calico/node:v1.0.0`

What we missed is: `-v /etc/tls/etcdclientpem:/etc/calico/certs/key.pem`

## Error log
```
goroutine 1 [running]:
panic(0x12f08c0, 0xc4203fb1b0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
main.main()
	/go/src/github.com/projectcalico/calico-containers/calico_node/startup.go:42 +0x50b
Calico node failed to start
time="2016-12-30T08:21:54Z" level=info msg="Creating Calico client" 
time="2016-12-30T08:21:54Z" level=info msg="Loading config from environment" 
panic: Error creating client: open /etc/calico/certs/key.pem: no such file or directory
```

### PR DESCRIPTION
**Pull request by gunjan5**
Fixes #1433

### CODE REVIEW

**Comment by robbrockbank** [comment]
Nope.

The problem is that the *filename* should be specified NOT the directory.  Perhaps we could add logic to check that it's not a directory.

**Comment by gunjan5** [comment]
Yup makes sense, but I don't think we need a logic to check if it's not a directory since as you said the fields say the word `File` which is very clear.

**Comment by gunjan5** [comment]
Reopening this as per the last 2 comments in the issue #1433

**Comment by gunjan5** [comment]
(needs re-testing once startup code is ported to Go)

**Comment by gunjan5** [comment]
Just tested this again since all the python code (startup) is replaced by Go.
I created all separate files for key, cert and CA first, that seemed to work, then I had a combined file for key and cert into a single file and that seemed to work as well:

```yaml
cat cfg.yaml
apiVersion: v1
kind: calicoApiConfig
metadata:
spec:
  datastoreType: etcdv2
  etcdEndpoints: https://127.0.0.1:2379
  etcdKeyFile: /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/client-key-cert.pem
  etcdCertFile: /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/client-key-cert.pem
  etcdCACertFile: /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/ca.pem
```

``` sudo ./dist/calicoctl -l debug node run --config=cfg.yaml
INFO[0000] Log level set to debug
INFO[0000] Loading config from JSON or YAML data
INFO[0000] Datastore type: etcdv2
INFO[0000] Include docker networking volume mounts
INFO[0000] Running in container

Running the following command to start calico-node:

docker run --net=host --privileged --name=calico-node -d --restart=always -e NODENAME=node-01 -e NO_DEFAULT_POOLS= -e CALICO_LIBNETWORK_ENABLED=true -e ETCD_SCHEME= -e ETCD_KEY_FILE=/etc/calico/certs/key.pem -e CALICO_NETWORKING_BACKEND=bird -e CALICO_LIBNETWORK_IFPREFIX=cali -e ETCD_ENDPOINTS=https://127.0.0.1:2379 -e ETCD_AUTHORITY= -e ETCD_CA_CERT_FILE=/etc/calico/certs/ca_cert.crt -e ETCD_CERT_FILE=/etc/calico/certs/cert.crt -v /var/log/calico:/var/log/calico -v /var/run/calico:/var/run/calico -v /lib/modules:/lib/modules -v /run/docker/plugins:/run/docker/plugins -v /var/run/docker.sock:/var/run/docker.sock -v /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/ca.pem:/etc/calico/certs/ca_cert.crt -v /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/client-key-cert.pem:/etc/calico/certs/key.pem -v /home/vagrant/go/src/github.com/projectcalico/calicoctl/certs/client-key-cert.pem:/etc/calico/certs/cert.crt quay.io/calico/node:latest
```

File: calicoctl/commands/node/run.go [hunk 1]
```diff
@@ -229,17 +229,23 @@ Description:
...
+	// Create a struct for volumes to mount.
+	type vol struct {
+		hostPath      string
+		containerPath string
 	}
 
+	// vols is a slice of read only volume bindings.
+	var vols []vol
+
+	vols = append(vols, vol{hostPath: logDir, containerPath: "/var/log/calico"},
```
Comment (robbrockbank) [inline calicoctl/commands/node/run.go]: Marginally easier to just vols as a slice containing these three entries.

`vols := []vol{{hostPath: logDir, containerPath: "/var/log/calico"}, {...}, {...}}`

**Comment by robbrockbank** [comment]
One minor comment, but LGTM.

File: calicoctl/commands/node/run.go (hunk 1 above)
Comment (gunjan5) [inline calicoctl/commands/node/run.go]: done!

**Pull request merged**